snapcraft/internal/pluginhandler/__init__.py

```diff
@@ -1071,3 +1087,40 @@ def check_for_collisions(parts):
         # And add our files to the list
         parts_files[part.name] = {'files': part_files,
                                   'installdir': part.installdir}
+
+
+def _get_includes(fileset):
+    return [x for x in fileset if x[0] != '-']
+
+
+def _get_excludes(fileset):
+    return [x[1:] for x in fileset if x[0] == '-']
+
+
+def _combine_filesets(fileset_1, fileset_2):
+    """Combine filesets if the first is an explicit or implicit wildcard."""
+
+    f1_excludes = set(_get_excludes(fileset_1))
+    f2_includes = set(_get_includes(fileset_2))
+
+    contradicting_fileset = set.intersection(f1_excludes, f2_includes)
+
+    if contradicting_fileset:
+        raise EnvironmentError('File conflicts: {key!r}'.format(
+                               key=contradicting_fileset))
```

Comment:
1) This isn't really an `EnvironmentError`, is it? Should we make a new Snapcraft error class for this?
2) This error is going to look like "File conflicts: {a, b}." I don't know how much that will inform a brand new snapcraft user, and we have the information at hand (in the caller if necessary) to make this a more informative/helpful message.